File: data-skill/templates/r/plot_dag.R
#!/usr/bin/env Rscript
# DAG (Directed Acyclic Graph) Template
# Purpose: Visualize causal relationships and identify confounders

library(ggplot2)
library(dagitty)
library(ggdag)

# ============================================================================
# USER INPUT: Define your DAG
# ============================================================================
# Format: "parent -> child; parent -> child"
dag_definition <- "dag {
    treatment -> outcome
    confounder -> treatment
    confounder -> outcome
    mediator -> outcome
    treatment -> mediator
    collider -> outcome
}"

# Alternative: Build programmatically
# dag <- dagitty::dagitty("dag {
#     treatment -> outcome
#     confounder -> treatment
#     confounder -> outcome
# }")

# ============================================================================
# PLOT DAG
# ============================================================================
dag <- dagitty::dagitty(dag_definition)

# Basic plot
ggdag(dag) +
  theme_dag() +
  labs(title = "Causal DAG")

# Colored by variable type
ggdag(dag) +
  theme_dag() +
  labs(title = "Causal DAG") +
  scale_gdag_d(fill = "variable")  # Color by variable type

# With adjustment sets highlighted
adjustments <- adjustmentSets(dag, exposure = "treatment", outcome = "outcome")
ggdag(dag) +
  theme_dag() +
  labs(title = "Causal DAG with Adjustment Sets") +
  geom_dag_node(aes(color = after_adjustment))

# ============================================================================
# OUTPUT ANALYSIS
# ============================================================================
cat("\n========================================\n")
cat("DAG ANALYSIS RESULTS\n")
cat("========================================\n\n")

# Show parents and children
cat("Parents of 'treatment':\n")
print(parents(dag, "treatment"))

cat("\nChildren of 'treatment':\n")
print(children(dag, "treatment"))

cat("\nAdjustment sets for 'treatment' -> 'outcome':\n")
print(adjustments)

cat("\nPaths from treatment to outcome:\n")
print(paths(dag, from = "treatment", to = "outcome"))

cat("\n========================================\n")

# Save plot
ggsave("dag_plot.png", width = 8, height = 6, dpi = 300)
cat("DAG plot saved to: dag_plot.png\n")

File: data-skill/templates/python/plot_dag.py
#!/usr/bin/env python3
# DAG (Directed Acyclic Graph) Template
# Purpose: Visualize causal relationships and identify confounders

import networkx as nx
import matplotlib.pyplot as plt
import pandas as pd

# ============================================================================
# USER INPUT: Define your DAG edges
# ============================================================================
# Format: (parent, child) tuples
dag_edges = [
    ("treatment", "outcome"),
    ("confounder", "treatment"),
    ("confounder", "outcome"),
    ("mediator", "outcome"),
    ("treatment", "mediator"),
    ("collider", "outcome")  # WARNING: Don't condition on collider!
]

# ============================================================================
# CREATE DAG
# ============================================================================
dag = nx.DiGraph()
dag.add_edges_from(dag_edges)

# Check for cycles
if not nx.is_directed_acyclic_graph(dag):
    print("Warning: Graph contains cycles - not a valid DAG!")

# ============================================================================
# PLOT DAG
# ============================================================================
plt.figure(figsize=(8, 6))

# Layout options
pos = nx.spring_layout(dag)  # Force-directed
# pos = nx.circular_layout(dag)  # Circular
# pos = nx.shell_layout(dag)      # Hierarchical

nx.draw(dag, pos,
        with_labels=True,
        arrows=True,
        node_size=2000,
        node_color='lightblue',
        edge_color='gray',
        font_size=10,
        arrowsize=20)

plt.title("Causal DAG")
plt.axis('off')
plt.tight_layout()

# Save plot
plt.savefig("dag_plot.png", dpi=300, bbox_inches='tight')
print("DAG plot saved to: dag_plot.png")

# ============================================================================
# OUTPUT ANALYSIS
# ============================================================================
print("\n" + "="*40)
print("DAG ANALYSIS RESULTS")
print("="*40 + "\n")

# Find parents and children
print("Parents of 'treatment':")
for parent in dag.predecessors('treatment'):
    print(f"  - {parent}")

print("\nChildren of 'treatment':")
for child in dag.successors('treatment'):
    print(f"  - {child}")

# Find all paths between two nodes
try:
    paths = list(nx.all_simple_paths(dag, 'treatment', 'outcome', cutoff=5))
    print(f"\nPaths from 'treatment' to 'outcome' (max 5 shown):")
    for i, path in enumerate(paths[:5], 1):
        print(f"  Path {i}: {' -> '.join(path)}")
except:
    print("\nNo paths found between 'treatment' and 'outcome'")

# Find adjustment sets (manual approach)
# In R use dagitty::adjustmentSets()
print("\n" + "="*40)
print("ADJUSTMENT SETS")
print("="*40 + "\n")
print("Note: For full adjustment sets, use R's dagitty package")
print("Minimum adjustment set for treatment -> outcome:")

# Find minimal adjustment set
confounders = []
for parent in dag.predecessors('treatment'):
    if parent != 'outcome':  # Not a mediator
        if parent in dag.predecessors('outcome'):
            confounders.append(parent)

if confounders:
    print(f"  {', '.join(confounders)}")
else:
    print("  (none needed - no common confounders)")

print("\n" + "="*40)
